File: kmeans.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <float.h>

#define TOLERANCE 0.01f
#define LINE_SIZE 1000

typedef struct{
  float **points;
  float *centroid;
}cluster;

cluster *clusters;

float **AllPoints;
int Dimensions;

//------------------------------------------------------------------------------
// Function declaration
//------------------------------------------------------------------------------
void init(int k, int n);
void recalcCentroids(int k, int n);
double sse(int k, int n);
void assignClusters(int k, int n);
void cluster_(int k, int n);
int read();
double euclid(float *point1, float *point2);
void normalizeAndRemoveOutlier(int n);


/*******************************************************************************
* Find euclidean distance between two points
* TODO make sure Dimensions gets set when file read in
*******************************************************************************/
double euclid(float *point1, float *point2){
  float sum = 0;
  for(int x = 0; x < Dimensions; x++){
    sum += pow(point1[x] - point2[x], 2);
  }
  return sqrt(sum);
}


/*******************************************************************************
* reads in the csv file and allocates needed memory to store it
*******************************************************************************/
int read(){
  FILE *doc;
  int ctr = 0;
  Dimensions = 1;
  int lines = 1000;
  AllPoints = malloc(sizeof(float**) * lines);

  if ((doc = fopen("proj02data.csv", "r")) == NULL){
    printf("Couldn't open file");
    exit(1);
  }
  char currentLine[LINE_SIZE];
  fgets(currentLine,LINE_SIZE,doc); //skip first line


  //figure out how many Dimensions there are
  int myBool = 0;
  while(fgets(currentLine,LINE_SIZE,doc) != NULL){
    //find the number of Dimensions first time through
    if(!myBool){
      char copy[LINE_SIZE];
      strcpy(copy, currentLine);
      strtok(copy, ",");
      while(strtok(NULL, ",") != NULL) Dimensions++;
    }
    myBool = 1;

    //reallocate memory if there are more lines than expected
    if(ctr >= lines - 1){
      lines *= 2;
      AllPoints = realloc(AllPoints, sizeof(float**) * lines); //get more memory
    }
    AllPoints[ctr] = malloc(sizeof(float*) * Dimensions);
    AllPoints[ctr][0] = atof(strtok(currentLine, ","));
    #ifdef DEBUG
      printf("AllPoints: %f\t", AllPoints[ctr][0]);
    #endif
    for(int x = 1; x < Dimensions; x++){
      AllPoints[ctr][x] = atof(strtok(NULL, ","));
      #ifdef DEBUG
        printf("AllPoints: %f\t", AllPoints[ctr][x]);
      #endif
    }
    #ifdef DEBUG
      printf("\n");
    #endif
    ctr++;
  }
  AllPoints = realloc(AllPoints, sizeof(float**) * ctr); //shrink to exact value needed
  return ctr;
}


/*******************************************************************************
* the algorithm phase, pretty much ties everything together
*******************************************************************************/
void cluster_(int k, int n){
  clusters = malloc(sizeof(cluster) * k);
  for(int x = 0; x < k; x++){
    clusters[x].points = malloc(sizeof(float**) *n);
    clusters[x].centroid = malloc(sizeof(float*) * Dimensions);
  }
  normalizeAndRemoveOutlier(n);
  init(k, n);
  assignClusters(k, n);
  double errThis = sse(k, n);
  double errLast = errThis + 1;
  while(errLast - errThis > TOLERANCE){
    assignClusters(k,n);
    recalcCentroids(k,n);
    errLast = errThis;
    errThis = sse(k,n);
  }
  //TODO finish
}


/*******************************************************************************
* assigns each point to a cluster given its distance to the centroid
* leaves random NULL values in some of the cluster points that will have to be
* ignored later on
*******************************************************************************/
void assignClusters(int k, int n){
  for(int x = 0; x < k; x++){
    for(int y = 0; y < n; y++){
      clusters[x].points[y] = NULL; //clear out any former clusters
    }
  }
  for(int x = 0; x < n; x++){
    int y1 = -1;
    double lowestDistance = DBL_MAX;
    double currentDistance;

    for(int y = 0; y < k; y++){
      currentDistance = euclid(AllPoints[x], clusters[y].centroid);
      if(currentDistance < lowestDistance){
        lowestDistance = currentDistance;
        y1 = y;
      }
    }
    clusters[y1].points[x] = AllPoints[x];
  }
  #ifdef DEBUG_LOTS
    for(int x = 0; x < k; x++){
      for(int y = 0; y < n; y++){
        printf("cluster %i:\t", x);
        for(int z = 0; z < Dimensions; z++){
          if(clusters[x].points[y] != NULL) printf("%f\t",clusters[x].points[y][z]);
        }
        printf("\n");
      }
    }
  #endif
}


/*******************************************************************************
* randomly assigns centroids to given clusters
*******************************************************************************/
void init(int k, int n){
  srand(time(NULL));
  for(int x = 0; x < k; x++){
    for(int y = 0; y < Dimensions; y++){
      clusters[x].centroid[y] = AllPoints[rand() % n][y];
      #ifdef DEBUG
        printf("centroid: %f\t", clusters[x].centroid[y]);
      #endif
    }
    #ifdef DEBUG
      printf("\n");
    #endif
  }
}


/*******************************************************************************
* calculate sum squared error between cluster points and their centroid
*******************************************************************************/
double sse(int k, int n){
  double error = 0;
  for(int x = 0; x < n; x++){
    for(int y = 0; y < k; y++){
      if(clusters[y].points[x] != NULL){
        error += euclid(clusters[y].points[x], clusters[y].centroid);
      }
    }
  }
  error = pow(error,2);
  return error;
}


/*******************************************************************************
* re-calculate centroid points for clusters
*******************************************************************************/
void recalcCentroids(int k, int n){
  for(int x = 0; x < k; x++){ //for each cluster
    int ctr = 0;
    float *temp = calloc(Dimensions, sizeof(float*));
    for(int y = 0; y < n; y++){ //for each point
      if(clusters[x].points[y] != NULL){
        ctr++; //count the number of points in the cluster
        for(int z = 0;z < Dimensions; z++){ //for each variable
          temp[z] += clusters[x].points[y][z]; //add up all the variables
        }
      }
    }
    printf("Recalculated centroid for cluster %i:",x);
    for(int z = 0; z < Dimensions; z++){
      if(ctr != 0) clusters[x].centroid[z] = temp[z]/(float)ctr; //find the mean
      #ifdef DEBUG
        printf("%f\t",clusters[x].centroid[z]);
      #endif
    }
    #ifdef DEBUG
      printf("\n");
    #endif
  }
}


/*******************************************************************************
*   Normalize the data and removes outliers in data set
*   Uses core normalization so values don't lie perfect between 0-1, the majority
*   will lie between these values but a few may lie outside them
*******************************************************************************/
void normalizeAndRemoveOutlier(int n){
  float *temp = calloc(Dimensions, sizeof(float*));
  float *max = calloc(Dimensions, sizeof(float*));
  float *min = calloc(Dimensions, sizeof(float*));
  for(int x = 0; x < Dimensions; x++){
    for(int y = 0; y < n; y++){
      if(AllPoints[y][x] > max[x]) max[x] = AllPoints[y][x]; //find min value
      if(AllPoints[y][x] < min[x]) min[x] = AllPoints[y][x]; //find max value
      temp[x] += AllPoints[y][x]; //add up all values
    }
    temp[x] /= (float) n; //find the mean
  }

  for(int x = 0; x < Dimensions; x++){
    for(int y = 0; y < n; y++){
      if(AllPoints[y][x] > temp[x]*2){
        AllPoints[y][x] = temp[x]; //remove outlier by making it the mean
      }else if(AllPoints[y][x] < temp[x]/5){
        AllPoints[y][x] = temp[x]; //remove low outliers
      }
      AllPoints[y][x] = (AllPoints[y][x] - min[x]) / (max[x] - min[x]); // normalize
    }
  }

  #ifdef DEBUG
    for(int x = 0; x < n; x++){
        printf("All Points %i:", x);
      for(int y = 0; y < Dimensions; y++){
          printf(" %f\t", AllPoints[x][y]);
      }
        printf("\n");
    }
  #endif
}





/*******************************************************************************
* main function
*******************************************************************************/
int main(int argc, char *argv[]){
  int n = read();
  cluster_(4, n);
  return 0;
}

File: Kmeans.java
import java.io.*;
import java.util.ArrayList;

public class Kmeans{

	private final double TOLERANCE=.01;

	public static void main(String[] args){
		Kmeans km=new Kmeans();
		double[][] inst=km.read("proj02data.csv");
		int[] c=km.cluster(inst,4);
		for(int i=0; i<inst.length; i++)
			System.out.println(i+"\t"+c[i]);
	}

	public int[] cluster(double[][] inst, int k){
		int[] clusters=new int[inst.length];
		double[][] centroids=init(inst,k);
		double errThis=sse(inst,centroids,clusters), errLast=errThis+1;
		while(errLast-errThis>TOLERANCE){
			//reassign the clusters using assignClusters
      clusters = assignClusters(inst, centroids);
			//re-calculate the centroids
      centroids = recalcCentroids(inst, clusters, k);
			//re-calculate the error using sse
      errLast = errThis;
      errThis = sse(inst, centroids, clusters);
		}
		return clusters;
	}

	//finds initial clusters - no modifications necessary
	public double[][] init(double[][] inst, int k){
		int n=inst.length, d=inst[0].length;
		double[][] centroids=new double[k][d];
		double[][] extremes=new double[d][2];
		for(int i=0; i<d; i++)
			extremes[i][1]=Double.MAX_VALUE;
		for(int i=0; i<n; i++)
			for(int j=0; j<d; j++){
				extremes[j][0]=Math.max(extremes[j][0],inst[i][j]);
				extremes[j][1]=Math.min(extremes[j][1],inst[i][j]);
			}
		for(int i=0; i<k; i++)
			for(int j=0; j<d; j++)
				centroids[i][j]=Math.random()*(extremes[j][0]-extremes[j][1])+extremes[j][1];
		return centroids;
	}

	public int[] assignClusters(double[][] inst, double[][] centroids){
		int n=inst.length, d=inst[0].length, k=centroids.length;
		int[] rtn=new int[n];
    //TODO check if correct: rtn maps each inst to centroid?
		//for each instance
    for(int x = 0; x < inst.length; x++){
      double lowestDistance = Double.MAX_VALUE;
      double currentDistance;
      int chosenCluster = -1;
      for(int centroid = 0; centroid < k; centroid++){
        //calculate the distance to each of the different centroids
        currentDistance = euclid(inst[x], centroids[centroid]);
        if(currentDistance < lowestDistance){
          lowestDistance = currentDistance;
          chosenCluster = centroid;
        }
      }
      //and assign it to the cluster with the lowest distance
      rtn[x] = chosenCluster;
    }
		return rtn;
	}

	public double[][] recalcCentroids(double[][] inst, int[] clusters, int k){
		int n=inst.length, d=inst[0].length;
		double[][] centroids=new double[k][d];
		int[] cnt=new int[k];
    //TODO
		//use cnt to count the number of instances in each cluster
		//for each cluster
		for(int x = 0; x < clusters.length; x++){
			//for each attribute in this cluster
			for(int y = 0; y < )
				//add the value of the attribute from each instance in the cluster
		}

		//calculate the averages by dividing each attribute total by the count
		//do this for each centroid, each attribute
		//be careful not to divide by zero - if a cluster is emply, skip it
		return centroids;
	}

	public double sse(double[][] inst, double[][] centroids, int[] clusters){
		int n=inst.length, d=inst[0].length, k=centroids.length;
		double sum=0;
    //TODO test if correct: check if this makes any sense at all
		//iterate through all instances
    for(int x = 0; x < n; x++){
      //iterate through all centroids
      for(int z = 0; z < k; z++){
        //if an instance is in the current cluster, add the euclidean distance
				//between them to the sum
        if(clusters[x] == z){ //if the current instance is mapped to the current centroid
          sum += euclid(inst[x], centroids[z]);
        }
      }
    }
		return sum;
	}

	private double euclid(double[] inst1, double[] inst2){
		double sum=0;
    //TODO test if correct
    for(int x = 0; x < inst1.length; x++){
      sum += Math.pow(inst1[x] - inst2[x], 2);
    }
		//calculate the euclidean distance between inst1 and inst2
		return Math.sqrt(sum);
	}

	//prints out a matrix - can be used for debugging - no modifications necessary
	public void printMatrix(double[][] mat){
		for(int i=0; i<mat.length; i++){
			for(int j=0; j<mat[i].length; j++)
				System.out.print(mat[i][j]+"\t");
			System.out.println();
		}
	}

	//reads in the file - no modifications necessary
	public double[][] read(String filename){
		double[][] rtn=null;
		try{
			BufferedReader br=new BufferedReader(new FileReader(filename));
			ArrayList<String> lst=new ArrayList<String>();
			br.readLine();//skip first line of file - headers
			String line="";
			while((line=br.readLine())!=null)
				lst.add(line); //read file into lst
			int n=lst.size(), d=lst.get(0).split(",").length;
			rtn=new double[n][d]; //# rows and columns from lst
			for(int i=0; i<n; i++){
				String[] parts=lst.get(i).split(","); // get each object in column
				for(int j=0; j<d; j++)
					rtn[i][j]=Double.parseDouble(parts[j]); //put each object into return
			}
			br.close();
		}catch(IOException e){System.out.println(e.toString());}
		return rtn;
	}

}
